### PR TITLE
feat(forge): get-solc command

### LINKED ISSUE #7656: feat(forge): add command to check solc autodetected version
### Component

Forge

### Describe the feature you would like

Foundry has a config which is auto detect solc version. It defaults to true and when you type forge build, it detects the appropriate newest allowed version and uses that.

I propose you add a "forge detect-sol" command which spits out the detected version.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by kamuik16**
Closes #7656 

## Motivation

Introduces `forge get-solc <path>` to get the solc version for that particular path.

This command can be useful to check compatible solc for different folders in a large project.

```bash
Solc 0.8.22
```
In case of multiple solcs:
```bash
Solcs 0.8.21, 0.8.22, 0.8.24
```
It works even if you give a file path like `src/test/Test.sol`.

### CODE REVIEW

File: crates/forge/bin/cmd/getsolc.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+pub struct GetSolcArgs {
+    /// The path of the project to get solc for.
+    #[arg(value_hint = ValueHint::DirPath, default_value = ".", value_name = "PATH")]
+    path: PathBuf,
+}
+
+impl GetSolcArgs {
+    pub fn run(self) -> Result<()> {
+        let GetSolcArgs { path } = self;
+        get_solc_from_path(path)?;
```
Comment (mattsse) [inline crates/forge/bin/cmd/getsolc.rs:18]: this should load the `Config` using the the path.
from which we can get the ProjectPaths

**Review by mattsse — CHANGES_REQUESTED** [review]
I just remembered, there is `forge tree` which I believe is very similar.

so do we even need this as an additional command @klkvr ?

File: crates/forge/bin/cmd/getsolc.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+use foundry_compilers::{artifacts::Source, utils, Graph, Project};
+use std::path::PathBuf;
+use yansi::Paint;
+
+/// CLI arguments for `forge get-solc`.
+#[derive(Clone, Debug, Parser)]
+pub struct GetSolcArgs {
+    /// The path of the project to get solc for.
+    #[arg(value_hint = ValueHint::DirPath, default_value = ".", value_name = "PATH")]
+    path: PathBuf,
```
Comment (klkvr) [inline crates/forge/bin/cmd/getsolc.rs:12]: can we rename this to `root: PathBuf` for consistency?

**Review by klkvr — CHANGES_REQUESTED** [review]
I think this is useful as this can give an insight into our version-resolving algorithm. It's not same as `forge tree` I believe as `forge tree` just prints files along with version requirements without triggering the algo.

This should use project config to resolve `ProjectPathsConfig` as @mattsse pointed out

File: crates/forge/bin/cmd/getsolc.rs (hunk 1 above)
Comment (kamuik16) [reply crates/forge/bin/cmd/getsolc.rs]: how can I get this? also I think resolve_sources adds the sources if they are not in the project.

Comment (kamuik16) [reply crates/forge/bin/cmd/getsolc.rs]: > this should load the `Config` using the the path. from which we can get the ProjectPaths

i did this but now if you specify a path for a folder which uses a single solc, the output is always something like this `Solcs 0.8.19, 0.8.21, 0.8.24` no matter the given path, which wasn't the case earlier.

**Comment by kamuik16** [comment]
@mattsse leave your review when you're free. ref https://github.com/foundry-rs/foundry/pull/7696#discussion_r1569968048. Thanks!

File: crates/forge/bin/cmd/getsolc.rs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        let graph = Graph::resolve(&config.project_paths())?;
+        let (versions, _) = graph.into_sources_by_version(project.offline)?;
+        let sources_by_version = versions.get(&project)?;
+        let versions: Vec<_> =
+            sources_by_version.keys().map(|solc| solc.version_short().unwrap()).collect();
+        if versions.len() == 1 {
+            let version = &versions[0];
+            println!(
+                "{} {}.{}.{}",
+                Paint::green("Solc"),
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/getsolc.rs:30]: yansi has been updated with a breaking API change, please update this code

File: crates/forge/bin/cmd/getsolc.rs [hunk 4]
```diff
@@ -0,0 +1,45 @@
+use clap::{Parser, ValueHint};
+use eyre::Result;
+use foundry_compilers::Graph;
+use foundry_config::Config;
+use std::path::PathBuf;
+use yansi::Paint;
+
+/// CLI arguments for `forge get-solc`.
+#[derive(Clone, Debug, Parser)]
+pub struct GetSolcArgs {
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/getsolc.rs:10]: can this be folded into another command? Feels a bit weird that's standalone but not a big deal.

**Pull request closed without merging**